### PR TITLE
feat(components): combobox new component

### LINKED ISSUE #824: [combobox] create a new component

### PR DESCRIPTION
**Pull request by yinonov**
Closes #824

- [x] add support for popup width (or more specifically its assigned node)
- [ ] enable text field icon trailing, or compose a dedicated text field with alternate style (can be customized in element definition). or just use plain input in template

### CODE REVIEW

File: libs/components/src/lib/option/option.scss [hunk 1]
```diff
@@ -30,7 +30,6 @@
 	box-sizing: border-box;
 	align-items: center;
 	background-color: var(#{appearance.get-appearance-token(fill)});
-	border-radius: 6px;
```
Comment (yinonov) [inline libs/components/src/lib/option/option.scss:33]: requested by design?

File: libs/components/src/lib/combobox/ui.test.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+test('should show the component', async ({ page }: { page: Page }) => {
+	const template = extractHTMLBlocksFromReadme(
+		path.join(new URL('.', import.meta.url).pathname, 'README.md')
+	).reduce(
+		(htmlString: string, block: string) =>
+			`${htmlString} <div style="margin: 5px;">${block}</div>`,
+		''
+	);
+
```
Comment (yinonov) [inline libs/components/src/lib/combobox/ui.test.ts:20]: set the viewport width to not waste extra space and a large image

File: libs/components/src/lib/option/option.scss (hunk 1 above)
Comment (rinaok) [reply libs/components/src/lib/option/option.scss:33]: @yinonov IDK... It is your change: https://github.com/Vonage/vivid-3/pull/825/commits/418d54dbe942418ac7a935e76efba7b6bc3a87f4
I can remove it if not needed.

File: libs/components/src/lib/combobox/combobox.scss [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	padding: 4px;
+	background-color: var(#{appearance.get-appearance-token(fill)});
+	border-radius: 6px;
+	box-shadow: inset 0 0 0 1px var(#{appearance.get-appearance-token(outline)});
+	contain: paint;
+}
+
+.popup {
+	--_popup-width: 209px;
+}
```
Comment (rinaok) [inline libs/components/src/lib/combobox/combobox.scss:47]: @rachelbt How I make the popup be the same width as the combobox?

File: libs/components/src/lib/combobox/combobox.scss [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+	padding: 4px;
+	background-color: var(#{appearance.get-appearance-token(fill)});
+	border-radius: 6px;
+	box-shadow: inset 0 0 0 1px var(#{appearance.get-appearance-token(outline)});
+	contain: paint;
+}
+
+.popup {
+	--_popup-width: 209px;
```
[the remark this answers is not in the document]
Comment (rinaok) [reply libs/components/src/lib/combobox/combobox.scss]: @rachelbt Tried that but it didn't work, that's why I put 209px

File: libs/components/src/lib/popup/popup.scss [hunk 5]
```diff
@@ -1,26 +1,31 @@
...
 	}
 }
 
 .popup-wrapper {
-	position: fixed;
+	&:not(.absolute) {
+		position: fixed;
+	}
+	&.absolute {
+		position: absolute;
```
Comment (rachelbt) [inline libs/components/src/lib/popup/popup.scss:20]:
```suggestion
		position: absolute;
		inline-size: var(#{variables.$popup-width}
```

File: libs/components/src/lib/combobox/combobox.scss [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+);
+@use "../../../../shared/src/lib/sass/mixins/connotation" as connotation;
+
+@use "../../../../shared/src/lib/sass/mixins/appearance/config" as appearance-config with (
+	$appearances: fieldset ghost,
+	$states: idle disabled,
+	$default: fieldset
+);
+@use "../../../../shared/src/lib/sass/mixins/appearance" as appearance;
+
```
Comment (rachelbt) [inline libs/components/src/lib/combobox/combobox.scss:15]:
```suggestion
:host {
  position: relative;
  }
```

**Comment by rinaok** [comment]
> Closes #824
> 
> * [x]  add support for popup width (or more specifically its assigned node)
> * [ ]  enable text field icon trailing, or compose a dedicated text field with alternate style (can be customized in element definition). or just use plain input in template

@yinonov Are you sure we need icon inside the combobox? I think that chevron is enough .

File: libs/components/src/lib/combobox/README.md [hunk 7]
```diff
@@ -66,10 +66,10 @@ Add a `placeholder` attribute to add placeholder text to the input.
 ```
 ### Autocomplete
 
-- Type: `'inline'` | `'list'` | `'both'` | `undefined`
+- Type: `'inline'` | `'list'` | `'both'`
```
Comment (yinonov) [inline libs/components/src/lib/combobox/README.md:69]: good fix. should we be adding `none`? https://github.com/microsoft/fast/tree/6ebb5ec7bdb7666f49ddc524bc56f70a2e8aa574/packages/web-components/fast-foundation/src/combobox#variables

**Review by rachelbt — COMMENTED** [review]
when opening the listbox - should the user navigate with the up & down arrows?

**Comment by rinaok** [comment]
> when opening the listbox - shouldn't the user navigate with the up & down arrows?

@rachelbt Yes I know, I'm working on it

**Comment by rinaok** [comment]
> > when opening the listbox - shouldn't the user navigate with the up & down arrows?
> 
> @rachelbt Yes I know, I'm working on it

@rachelbt It's working now

File: libs/components/src/lib/combobox/README.md (hunk 7 above)
Comment (yinonov) [reply libs/components/src/lib/combobox/README.md:69]: It's a native spec. not our call

**Review by rinaok — APPROVED** [review]

**Review by yinonov — COMMENTED** [review]
how does triggering the focus on an option work? on `keypress`? would it help to delegate focus in popup definition? so that focus on host will automatically decide which option to focus on (instead of manual)

**Comment by rinaok** [comment]
> how does triggering the focus on an option work? on `keypress`? would it help to delegate focus in popup definition? so that focus on host will automatically decide which option to focus on (instead of manual)

@yinonov You want us to remove all Fast's key handlers from the combobox template and that the popup will handle it?

**Comment by yinonov** [comment]
> rs from the combobox template and that the po

no! so it's a hindered knowledge to this scope? handled by fast? I'd still add tests to it. WDYT?

**Comment by rinaok** [comment]
> 

@yinonov Yes, all the behaviour is handled by Fast

**Review by yinonov — COMMENTED** [review]
why does the `open` section rendered snippet autofocus? I didn't see any explicit code for it (it might be intentional. just making sure)

image

File: change/@vonage-vivid-8865978f-8468-4691-a216-12a13c91a92f.json [hunk 8]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (yinonov) [inline change/@vonage-vivid-8865978f-8468-4691-a216-12a13c91a92f.json:1]: is this redundant?

File: libs/components/src/lib/combobox/README.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+```js
+<script type="module">
+  import '@vonage/vivid/combobox';
+</script>
+```
+
+## Slots
+
+### Option
```
Comment (yinonov) [inline libs/components/src/lib/combobox/README.md:13]: it should be `default`. as in
https://github.com/Vonage/vivid-3/blob/9bafb96d81556e9cff522046acd2ca166da29b01/libs/components/src/lib/menu/README.md#L78-L81

Comment (rinaok) [reply libs/components/src/lib/combobox/README.md]: @yinonov updated but we need to update it also in listbox readme

File: libs/components/src/lib/combobox/README.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+- Type: `string` | `undefined`
+- Default: `undefined`
+
+Set the `value` attribute to set the default value for the input field. Setting the property on the element will not change the default value, but will change the value shown in the view as well as the submitted value in a form (imitating the native behavior).
+
+```html preview
+<vwc-combobox value="value">
+ <vwc-option text="First Option"></vwc-option>
+ <vwc-option text="Second Option"></vwc-option>
+</vwc-combobox>
```
Comment (olaf-k) [inline libs/components/src/lib/combobox/README.md:51]: this example does not demonstrate the `value` usage.
```suggestion
<vwc-combobox value="First Option">
 <vwc-option text="First Option"></vwc-option>
 <vwc-option text="Second Option"></vwc-option>
</vwc-combobox>
```

also, and this is beyond the scope of your PR, but why are `option` text and value the same thing? they should be be two separate props/attrs

**Review by olaf-k — CHANGES_REQUESTED** [review]

**Comment by olaf-k** [comment]
clicking on the combobox button opens the list of options, but clicking again does not close it.

**Comment by olaf-k** [comment]
if you have a page with a combobox followed by a listbox, the combobox list appears under the listbox when opened:
chrome_DdqtA2DSml

File: libs/components/src/lib/combobox/README.md (hunk 10 above)
Comment (yinonov) [reply libs/components/src/lib/combobox/README.md:51]: I think options here should have a matching value. Don't they?

**Comment by yinonov** [comment]
U mean it doesn't go on top? Top-layer access issue. Referenced in our docs in multiple caveat sections

File: libs/components/src/lib/combobox/README.md (hunk 10 above)
Comment (olaf-k) [reply libs/components/src/lib/combobox/README.md:51]: > I think options here should have a matching value. Don't they?

the FAST and native implementations both distinguish the displayed text from the form value, I was expecting our `option` to behave in the same manner.

Comment (yinonov) [reply libs/components/src/lib/combobox/README.md:51]: of course. sorry, didn't catch ur previous comment. I mean combobox value should match an option value

File: libs/components/src/lib/popup/partials/variables.scss [hunk 11]
```diff
@@ -1 +1 @@
-$popup-width: --_popup-width;
+$popup-width: --popup-width;
```
Comment (rachelbt) [inline libs/components/src/lib/popup/partials/variables.scss]:
```suggestion
.popup {
	$popup-width: --_popup-width;
	#{$popup-width}: 50%;
}
```

File: libs/components/src/lib/combobox/combobox.scss [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+	padding: 4px;
+	background-color: var(#{appearance.get-appearance-token(fill)});
+	border-radius: 6px;
+	box-shadow: inset 0 0 0 1px var(#{appearance.get-appearance-token(outline)});
+	contain: paint;
+}
+
+.popup {
+	--popup-width: 100%;
```
Comment (rachelbt) [inline libs/components/src/lib/combobox/combobox.scss]:
```suggestion
	$popup-width: --_popup-width;
	#{$popup-width}: 100%;
```

File: libs/components/src/lib/popup/partials/variables.scss (hunk 11 above)
Comment (rachelbt) [inline libs/components/src/lib/popup/partials/variables.scss]:
```suggestion
$popup-width: --_popup-width;
```

**Comment by rinaok** [comment]
> 

@olaf-k @yinonov It's the same behaviour as in FAST https://explore.fast.design/components/fast-combobox, do you think it's a bug?

**Comment by olaf-k** [comment]
> @olaf-k @yinonov It's the same behaviour as in FAST [explore.fast.design/components/fast-combobox](https://explore.fast.design/components/fast-combobox), do you think it's a bug?

good point, I checked and [they already have an issue on their side about it](https://github.com/microsoft/fast/issues/5721).

**Comment by yinonov** [comment]
> why does the `open` section rendered snippet autofocus? I didn't see any explicit code for it (it might be intentional. just making sure)
> 
> image

https://github.com/microsoft/fast/blob/6ebb5ec7bdb7666f49ddc524bc56f70a2e8aa574/packages/web-components/fast-foundation/src/combobox/combobox.ts#L336-L352

**Comment by rinaok** [comment]
> U mean it doesn't go on top? Top-layer access issue. Referenced in our docs in multiple caveat sections (e.g. [tooltip#caveat](https://vivid.deno.dev/components/tooltip#caveat)). perhaps a similar caveat mention should be added for combobox

@yinonov Added Caveat

**Review by rinaok — APPROVED** [review]

**Review by rinaok — DISMISSED** [review]

File: libs/components/src/lib/combobox/README.md [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+```
+
+## Caveat
+
+Document elements display precedence is formed by the imaginary z-axis [stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context), commonly by order of which elements are rendered and special properties (e.g. _z-index_).
+Combobox component is a low level element, unaware of its document context, but is, in most cases, required to overlay on top of all elements.
+
+A common practice used in apps / frameworks to promote a Combobox component to top other elements z-axis, is to utilise a service that dynamically appends a Combobox component to the end of the body element, when called for.
```
Comment (olaf-k) [inline libs/components/src/lib/combobox/README.md:143]: hmm. this probably is a valid advice for modal components that appear on their own but this is hardly applicable to form components such as the combobox. I'm ok to leave the widget as it is until someone complains but we need to be prepared for another solution than "dynamically append it to body".

**Review by olaf-k — CHANGES_REQUESTED** [review]
I'll open an issue for this option.value thingy.

File: libs/components/src/lib/combobox/README.md (hunk 13 above)
Comment (rinaok) [reply libs/components/src/lib/combobox/README.md:143]: ok

**Review by rinaok — DISMISSED** [review]

**Review by olaf-k — APPROVED** [review]
can you put back the rest of the disclaimer?

```
## Caveat

Document elements display precedence is formed by the imaginary z-axis [stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context), commonly by order of which elements are rendered and special properties (e.g. _z-index_).
Combobox component is a low level element, unaware of its document context, but is, in most cases, required to overlay on top of all elements.
```

**Review by rinaok — APPROVED** [review]

File: libs/components/src/lib/combobox/combobox.scss [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+	padding: 4px;
+	background-color: var(#{appearance.get-appearance-token(fill)});
+	border-radius: 6px;
+	box-shadow: inset 0 0 0 1px var(#{appearance.get-appearance-token(outline)});
+	contain: paint;
+}
+
+.popup {
+	$popup-width: --_popup-width;
```
Comment (yinonov) [inline libs/components/src/lib/combobox/combobox.scss:50]: how did `--_popup-width` pass stylelint? it should have thrown an error exception.

extract to, and import from `partials/_variables.scss`

File: libs/components/src/lib/combobox/ui.test.ts (hunk 2 above)
Comment (yinonov) [reply libs/components/src/lib/combobox/ui.test.ts:20]: @rinaok

File: libs/components/src/lib/combobox/combobox.scss (hunk 14 above)
Comment (rinaok) [reply libs/components/src/lib/combobox/combobox.scss]: done

Comment (rinaok) [reply libs/components/src/lib/combobox/combobox.scss]: @rachelbt @yinonov I think that we should add default value for popup-width inside popup css, right?

Screenshot 2022-12-21 at 16 15 14

File: libs/components/src/lib/combobox/combobox.scss [hunk 15]
```diff
@@ -47,7 +48,5 @@
 }
 
 .popup {
-	$popup-width: --_popup-width;
-	
-	#{$popup-width}: 100%;
+	#{popup-variables.$popup-width}: 100%;
```
Comment (yinonov) [inline libs/components/src/lib/combobox/combobox.scss:51]: can u explain this again? where is this variable used?

Comment (rinaok) [reply libs/components/src/lib/combobox/combobox.scss:51]: @yinonov When popup is strategy="absolute"

Comment (yinonov) [reply libs/components/src/lib/combobox/combobox.scss:51]: and?

Comment (yinonov) [reply libs/components/src/lib/combobox/combobox.scss:51]: never mind followed the code

**Review by rinaok — APPROVED** [review]

File: libs/components/src/lib/combobox/combobox.template.ts [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+import type { Combobox } from './combobox';
+import { focusTemplateFactory } from './../../shared/patterns/focus';
+
+/**
+ *
+ */
+function renderLabel() {
+	return html<Combobox>`
+	  <label for="control" class="label">
+		  ${x => x.label}
```
Comment (olaf-k) [inline libs/components/src/lib/combobox/combobox.template.ts:16]: mixed spaces with tabs here and in `comboboxTemplate` (should've been caught by linter though?)

Comment (yinonov) [reply libs/components/src/lib/combobox/combobox.template.ts:16]: something's off with eslint

**Pull request merged**